src/installer/iohandler.py

```diff
@@ -0,0 +1,100 @@
+"""Handling wheel installation I/O-related functionality."""
+
+import hashlib
+import os
+from io import RawIOBase
+
+from installer._compat.typing import TYPE_CHECKING
+
+from .records import Hash, Record
+
+if TYPE_CHECKING:
+    from ._compat.typing import FSPath
+    from typing import Dict, Tuple, Optional, Callable
+
+_BLOCK_SIZE = 4096
+_SCHEME_KEYS = ["platlib", "purelib"]
+
+
+def basic_copy_handler(
+    source,  # type: RawIOBase
+    dest,  # type: FSPath
+    hash_algorithm="sha256",  # type: Optional[str]
+):
+    # type: (...) -> Tuple[int, Hash]
+    """Copy a buffer and compute its hash and size.
+
+    Copies the source buffer to the specified destination. It also computes the hash
+    of the file to avoid traversing the source twice.
+
+    :param source: buffer holding the source data
+    :param dest: destination path
+    :param hash_algorithm: hashing algorithm
+
+    :return: size and ``Hash`` tuple of the copied buffer
+    """
+    assert hash_algorithm is not None  # to please mypy
```

Comment:
Why are we using an assert? mypy is right in calling it out, we shouldn't use an assert (which can be disabled with `-OO` :stuck_out_tongue:) just to please mypy. I think we should raise an exception.

Reply:
Why is `hash_algorithm` allowed to be None in the first place? What does it mean?

Reply:
That's API design I did not question it as I said the in general review commentary. I assumed there was a reason this could be `None` and given the assert, I suggested bumping it up to a proper exception.

But looking into it, `RECORD` does not have a hash algorithm and still needs to be copied. We should just skip the hash checking if `hash_algorithm` is `None` instead of exploding.
Breaking PEP 427, but happening in the wild, some backends are not generating hashes for `.pyc` files.
This seems to include setuptools, which is problematic. It is a bit late today, I will see if tomorrow I can look into it.